design/secure-credentials-store.md

```diff
@@ -21,51 +21,119 @@ Credentials include:
+    }
+
+    /**
+     * Retrieves a secret from vault. The value returned by the first vault provider that resolves
+     * the secret successfully (even to {@code null}) is returned;  if no vault provider resolves it,
+     * this method results into an empty {@link java.util.Optional}.
+     *
+     * @param context Context that contains info about current request, to provide e.g. realm separation
+     * @param vaultSecretId Identifier of the secret. It corresponds to the value entered by user
+     *        in the respective configuration, which in turn is obtained from the vault when storing
+     *        the secret.
+     * @param purpose Purpose of the secret, set by keycloak according to the caller context.
+     * @return The secret or {@code null} if the secret was successfully resolved, or
+     *         an empty {@link java.util.Optional}
+     */
+    Optional<byte[]> obtainSecret(KeycloakContext context, String vaultSecretId, String purpose);
+}
+```
+
+Vault configuration has to contain mapping between purpose and the allowed provider IDs sorted by priority to be used for this purpose, e.g. `realm-key => HSM, vault-1`, `smtp => vault-1`, `default => vault-2, HSM` would mean that:
+* realm keys should primarily be obtained from a provider with id `HSM` and then from a provider with ID `vault-1` (nothing else)
+* For SMTP secrets only `vault-1` should be consulted, and
+* For any other purpose, `vault-2` and `HSM` should be used to retrieve secrets in that order.
+
+### Vault Service Provider
+
+Individual vault services have to provide the secrets based on the context (e.g. realm) and the ID.
+They target the individual implementations.
+
+In Java pseudocode:
+
+```java
+public interface VaultServiceProvider {
+    /**
+     * Retrieves a secret from vault. The implementation should respect at least the realm ID
+     * to separate the secrets within the vault. If the secret is retrieved successfully,
+     * it is returned
+     *
+     * @param context Context that contains info about current request, to provide e.g. realm separation
+     * @param vaultSecretId Identifier of the secret. It corresponds to the value entered by user
+     *        in the respective configuration, which in turn is obtained from the vault when storing
+     *        the secret.
+     * @param purpose Purpose of the secret, set by keycloak according to the caller context.
+     * @return The secret or {@code null} if the secret was successfully resolved, or
+     *         an empty {@link java.util.Optional}
+     */
+    Optional<byte[]> obtainSecret(KeycloakContext context, String vaultSecretId);
+}
+```
+
+#### Default vault service providers to consider
+
+This is the list of vaults considered for a default implementations (TODO: to be shortlisted):
+
+* [Kubernetes](https://kubernetes.io/docs/concepts/configuration/secret/) / [OpenShift](https://docs.openshift.com/container-platform/4.1/nodes/pods/nodes-pods-secrets.html) secrets
```

Comment:
This one can be interpreted as a `paint text` vault. Kube can mount a secret as a key-value-based file.

Reply:
Good point. The Kubernetes provider would just read values from a file per secret, or from an environment value. That means the plain text is not needed, but rather the Kube/OpenShift one can just be used for testing purposes.